### PR TITLE
make classpath scans over circular symlinks predictable.

### PR DESCRIPTION
**Pull request by jct-amzn**
fixes https://github.com/google/guava/issues/2648 .

### CODE REVIEW

File: guava-tests/test/com/google/common/reflect/ClassPathTest.java [hunk 1]
```diff
@@ -153,6 +160,41 @@ public void testScan_classPathCycle() throws IOException {
...
+      Path right = createDirectory(root.resolve("right"));
+      createFile(right.resolve("another.txt"));
+
+      createSymbolicLink(left.resolve("sibling"), right);
+      createSymbolicLink(right.resolve("sibling"), left);
+
+      ClassPath.DefaultScanner scanner = new ClassPath.DefaultScanner();
+      scanner.scan(root.toFile(), loader);
+
+      assertEquals(ImmutableSet.of(
```
Comment (jbduncan) [inline guava-tests/test/com/google/common/reflect/ClassPathTest.java]: Lines 185-192 don't look like they're exactly in Google Java Style to me. If you haven't done so already, I suggest running [google-java-format](https://github.com/google/google-java-format) on this file.

File: guava/src/com/google/common/reflect/ClassPath.java [hunk 2]
```diff
@@ -478,10 +484,18 @@ protected void scanJarFile(ClassLoader classloader, JarFile file) {
 
     @Override
     protected void scanDirectory(ClassLoader classloader, File directory) throws IOException {
-      scanDirectory(directory, classloader, "");
+      scanDirectory(directory, classloader, "", new HashSet<>());
     }
 
-    private void scanDirectory(File directory, ClassLoader classloader, String packagePrefix)
+    /**
+     * Recursively scan the given directory, adding resources for each file encountered.  Symlinks which have already
```
Comment (jbduncan) [inline guava/src/com/google/common/reflect/ClassPath.java]: From my past experience, the Guava team prefer sentences be separated by one space only. I'd suggest running google-java-format on this file too, as that would (or at least should) fix the javadoc to meet their standards.

File: guava/src/com/google/common/reflect/ClassPath.java [hunk 3]
```diff
@@ -478,10 +484,18 @@ protected void scanJarFile(ClassLoader classloader, JarFile file) {
...
+     * been traversed in the current tree path will be skipped to eliminate cycles; otherwise symlinks are traversed.
+     * @param directory the root of the directory to scan
```
Comment (jbduncan) [inline guava/src/com/google/common/reflect/ClassPath.java]: A blank line between the last paragraph and the first `@param` is required AFAIK. Use google-java-format to fix this.

File: guava/src/com/google/common/reflect/ClassPath.java [hunk 4]
```diff
@@ -478,10 +484,18 @@ protected void scanJarFile(ClassLoader classloader, JarFile file) {
...
+     * @param classloader the classloader that includes resources found in {@code directory}
+     * @param packagePrefix resource path prefix inside {@code classloader} for any files found under {@code directory}
+     * @param currentPath canonical files already visited in the current directory tree path, for cycle elimination
+     */
+    private void scanDirectory(File directory, ClassLoader classloader, String packagePrefix, Set<File> currentPath)
```
Comment (jbduncan) [inline guava/src/com/google/common/reflect/ClassPath.java]: Is this method decl. <= 80 characters long? If not then running google-java-format should fix it. :)

**Comment by jct-amzn** [comment]
I signed it.  I should be covered by Amazon corporate CLA, though it's not appearing on my account for some reason.

**Comment by jbduncan** [comment]
@jct-amzn I think @googlebot hasn't recognised that you've signed the CLA? By my understanding, once you type the message

> I signed it!

then googlebot should automatically detect it (it's a GitHub service/bot, not a human user :smiley:) and it should tell you what you need to do next to change the red "cla: no" label into a green "cla: yes" label.

**Comment by jct-amzn** [comment]
Thanks for your review jbduncan.  I know googlebot's a bot -- I think the gears are turning on the corporate CLA.

My google account doesn't show that I'm part of the correct corporate CLA, so I think there's some problem on the backend, in the way the dots are all connected (github email -> google group -> CLA signatory).

Anyway it's just some one-time bookkeeping and I'm working on resolving it.

**Comment by jct-amzn** [comment]
I signed it!

**Comment by jct-amzn** [comment]
Ok, trying again.  Google CLA recognizes me now.

I signed it!

**Comment by jct-amzn** [comment]
checking back on the status of this.  the CLA verifier probably doesn't support private email addresses in the github account.

if the bot is using this app script to do the CLA check:
https://github.com/angular/google-cla-verifier-for-github/blob/master/cla-verifier.gs

then you can see that they're just pulling the "fake" email from the patch file:
https://github.com/angular/google-cla-verifier-for-github/blob/master/cla-verifier.gs#L210

so I think there is no way the bot can trace that back to my actual email address, which should be in their signatories spreadsheet.

**Comment by cpovirk** [comment]
Bulk closing all pull requests that are listed as needing CLA signing. If you'd like us to look at your pull request, you'll need to sign the CLA and report back here.

If this is a false positive, I apologize. Please reopen the pull request, and we'll have a look.

**Comment by willnorris** [comment]
CLA verified offline.  This is okay to merge.

**Comment by jct-amzn** [comment]
thanks Will.  Can you update the 'cla' label so the request doesn't get auto-closed again?

I'm going to work out a different arrangement at work so this isn't so complicated next time.

**Comment by cpovirk** [comment]
I've sent this to someone to review, so we'll take care of it before then next auto-closing. Sorry for the trouble.

**Comment by cpovirk** [comment]
The internal reviewer found one thing to tweak. You can have a look at 7c11b8d if you'd like. Thanks for the fix, and sorry for the CLA trouble.

**Comment by jbduncan** [comment]
Hi @cpovirk, as a external reviewer, I managed to find a couple of things in that commit which I think could be improved upon. I've shared my thoughts on those things in the comments below:

1. https://github.com/google/guava/commit/7c11b8d9220b710d5598b774e22cc774f68f40e7#commitcomment-20999208
2. https://github.com/google/guava/commit/7c11b8d9220b710d5598b774e22cc774f68f40e7#commitcomment-20999157
3. https://github.com/google/guava/commit/7c11b8d9220b710d5598b774e22cc774f68f40e7#commitcomment-20999165

**Comment by jct-amzn** [comment]
Thanks @cpovirk .  I'll keep AndroidIncompatible in mind for future commits.  No worries about the CLA, I know I was making it a challenge.

**Pull request closed without merging**